src/order/filter/germ.lean

```diff
@@ -0,0 +1,534 @@
+def map' {lc : filter γ} (F : (α → β) → (γ → δ)) (hF : (l.eventually_eq ⇒ lc.eventually_eq) F F) :
+  germ l β → germ lc δ :=
+quotient.map' F hF
+
+/-- A version of `quotient.lift_on'` using coercion instead of `quotient.mk'`. -/
+def lift_on {γ : Sort*} (f : germ l β) (F : (α → β) → γ) (hF : (l.eventually_eq ⇒ (=)) F F) : γ :=
+quotient.lift_on' f F hF
+
+@[simp] lemma map'_coe {lc : filter γ} (F : (α → β) → (γ → δ))
+  (hF : (l.eventually_eq ⇒ lc.eventually_eq) F F) (f : α → β) :
+  map' F hF f = F f :=
+rfl
+
+@[simp, norm_cast] lemma coe_eq : (f : germ l β) = g ↔ (f =ᶠ[l] g) := quotient.eq'
+
+alias coe_eq ↔ _ filter.eventually_eq.germ_eq
+
+/-- Lift a function `β → γ` to a function `germ l β → germ l γ`. -/
+def map (op : β → γ) : germ l β → germ l γ :=
+map' ((∘) op) $ λ f g H, H.mono $ λ x H, congr_arg op H
+
+@[simp] lemma map_coe (op : β → γ) (f : α → β) : map op (f : germ l β) = op ∘ f := rfl
+
+@[simp] lemma map_id : map id = (id : germ l β → germ l β) := by { ext ⟨f⟩, refl }
+
+lemma map_map (op₁ : γ → δ) (op₂ : β → γ) (f : germ l β) :
+  map op₁ (map op₂ f) = map (op₁ ∘ op₂) f :=
+induction_on f $ λ f, rfl
+
+/-- Lift a binary function `β → γ → δ` to a function `germ l β → germ l γ → germ l δ`. -/
+def map₂ (op : β → γ → δ) : germ l β → germ l γ → germ l δ :=
+quotient.map₂' (λ f g x, op (f x) (g x)) $ λ f f' Hf g g' Hg,
+Hg.mp $ Hf.mono $ λ x Hf Hg, by simp only [Hf, Hg]
+
+/-- Lift of `tendsto` to the space of germs. -/
+protected def tendsto (f : germ l β) (lb : filter β) : Prop :=
+lift_on f (λ f, tendsto f l lb) $ λ f g H, propext (tendsto_congr' H)
+
+@[simp, norm_cast] lemma coe_tendsto {f : α → β} {lb : filter β} :
+  (f : germ l β).tendsto lb ↔ tendsto f l lb :=
+iff.rfl
+
+alias coe_tendsto ↔ _ filter.tendsto.germ_tendsto
+
+/-- Composition of a germ and a `filter.germ.tendsto`. -/
+def comp_tendsto' (f : germ l β) {lc : filter γ} (g : germ lc α) (hg : g.tendsto l) :
+  germ lc β :=
+lift_on f (λ f, g.map f) $ λ f₁ f₂ hF, (induction_on g $ λ g hg, coe_eq.2 $ hg.eventually hF) hg
+
+@[simp] lemma coe_comp_tendsto' (f : α → β) {lc : filter γ} {g : germ lc α} (hg : g.tendsto l) :
+  (f : germ l β).comp_tendsto' g hg = g.map f :=
+rfl
+
+/-- Composition of a germ and a `filter.tendsto`. -/
+def comp_tendsto (f : germ l β) {lc : filter γ} (g : γ → α) (hg : tendsto g lc l) :
+  germ lc β :=
+f.comp_tendsto' _ hg.germ_tendsto
+
+@[simp] lemma coe_comp_tendsto (f : α → β) {lc : filter γ} {g : γ → α} (hg : tendsto g lc l) :
+  (f : germ l β).comp_tendsto g hg = f ∘ g :=
+rfl
+
+@[simp] lemma comp_tendsto'_coe (f : germ l β) {lc : filter γ} {g : γ → α} (hg : tendsto g lc l) :
+  f.comp_tendsto' _ hg.germ_tendsto = f.comp_tendsto g hg :=
+rfl
+
+@[simp, norm_cast] lemma const_inj (hl : l ≠ ⊥) {a b : β} : (↑a : germ l β) = ↑b ↔ a = b :=
+coe_eq.trans $ const_eventually_eq hl
+
+@[simp] lemma map_const (l : filter α) (a : β) (f : β → γ) :
+  (↑a : germ l β).map f = ↑(f a) :=
```

Comment:
Does it fail without the up arrows (here and below)?

Reply:
Yes because I defined only `has_lift_t` from `β` to `germ l β`. I don't think that silently upgrading from a constant to the corresponding constant germ is a good idea.